source/base/hdf5.cc

```diff
@@ -0,0 +1,1752 @@
+    //
+    // In the case of a variable length string, H5Awrite needs the address of a
+    // (char *). For this reason the std::string value has been copied to a C
+    // string.
+
+    hid_t  attr;
+    hid_t  aid;
+    hid_t  t_type;
+    herr_t ret;
+
+    // Reserve space for the string and the null terminator
+    char *c_string_value = (char *)malloc(sizeof(char) * (value.size() + 1));
+    strcpy(c_string_value, value.data());
+
+    /* Create a datatype to refer to. */
+    t_type = H5Tcopy(H5T_C_S1);
+    Assert(t_type >= 0, ExcInternalError());
+
+    // Python strings are encoded in UTF8
+    ret = H5Tset_cset(t_type, H5T_CSET_UTF8);
+    Assert(t_type >= 0, ExcInternalError());
+
+    ret = H5Tset_size(t_type, H5T_VARIABLE);
+    Assert(ret >= 0, ExcInternalError());
+
+    /*
+     * Create scalar attribute.
+     */
+    aid = H5Screate(H5S_SCALAR);
+    Assert(aid >= 0, ExcMessage("Error at H5Screate"));
+    attr = H5Acreate2(
+      *hdf5_reference, attr_name.data(), t_type, aid, H5P_DEFAULT, H5P_DEFAULT);
+    Assert(attr >= 0, ExcMessage("Error at H5Acreate2"));
+
+    /*
+     * Write scalar attribute.
+     */
+    ret = H5Awrite(attr, t_type, &c_string_value);
+    Assert(ret >= 0, ExcInternalError());
+
+    free(c_string_value);
+    ret = H5Sclose(aid);
+    Assert(ret >= 0, ExcMessage("Error at H5Sclose"));
+    ret = H5Aclose(attr);
+    Assert(ret >= 0, ExcMessage("Error at H5Aclose"));
+
+    (void)ret;
+  }
+
+
+
+  DataSet::DataSet(const std::string &name,
+                   const hid_t &      parent_group_id,
+                   const bool         mpi)
+    : HDF5Object(name, mpi)
+    , query_io_mode(false)
+    , io_mode(H5D_MPIO_NO_COLLECTIVE)
+    , local_no_collective_cause(H5D_MPIO_SET_INDEPENDENT)
+    , global_no_collective_cause(H5D_MPIO_SET_INDEPENDENT)
+  {
+    hdf5_reference = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+      // Relase the HDF5 resource
+      H5Dclose(*pointer);
+      delete pointer;
+    });
+    dataspace      = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+      // Relase the HDF5 resource
+      H5Sclose(*pointer);
+      delete pointer;
+    });
+
+    // rank_ret can take a negative value if the functions
+    // H5Sget_simple_extent_ndims and H5Sget_simple_extent_dims fail. rank is
+    // unsigned int therefore it can not be used to store the return value of
+    // H5Sget_simple_extent_ndims and H5Sget_simple_extent_dims
+    int rank_ret;
+
+    *hdf5_reference = H5Dopen2(parent_group_id, name.data(), H5P_DEFAULT);
+    Assert(*hdf5_reference >= 0, ExcMessage("Error at H5Dopen2"));
+    *dataspace = H5Dget_space(*hdf5_reference);
+    Assert(*dataspace >= 0, ExcMessage("Error at H5Dget_space"));
+    rank_ret = H5Sget_simple_extent_ndims(*dataspace);
+    Assert(rank_ret >= 0, ExcInternalError());
+    rank          = rank_ret;
+    hsize_t *dims = (hsize_t *)malloc(rank * sizeof(hsize_t));
```

Comment:
I think this could be done more easily if you did
```
  std::unique_ptr<hsize_t[]> dims = std_cxx14::make_unique<hsize_t[]>(rank);
```
Then you won't have to worry about the `free` statement below.

Reply:
I realized that it is easier to write the data directly to the vector dimensions. I was doing an unnecessary copy of the data. The new code:
``` c++
dimensions.resize(rank);
rank_ret = H5Sget_simple_extent_dims(*dataspace, dimensions.data(), NULL);
```